### DISCUSSION TITLE
Missing the missing UseExtension

### DISCUSSION THREAD

**Discussion by drewburlingame** [VerifyTests/Verify#782] [General]
We have several extension methods that help us format different types of output from our tests.  For example, we use Verify to generate "freeze files" for our arch-unit tests to capture violations that need to be fixed but we can't take them on just yet.  The freeze files have some specific conventions including using the extension "freeze.txt".  In another case, we generate markdown reports describing aspects of our code and they use several settings together.  Now that UseExtensions has been removed, the extension methods we use for formatting are missing a key aspect of the configuration and we have to correctly set the extension in all places the extension methods are used.  Any chance we could bring this feature back?

**Comment by SimonCropp** [comment]
can u share some runnable examples of your scenarios? perhaps push them up to a gh repo with the solution working against the older version of Verify

**Reply by drewburlingame** [reply]
Apologies for the delayed response.  I don't have runnable examples, but I think this will give you the concept.  We have these extension methods that we use to generate various types of reports on our code base.

```c#

public static class VerifySettingsExtensions
{
    public static SettingsTask AsFreezeFile(this SettingsTask settings) =>
        settings.UseExtension("freeze.txt");

    public static SettingsTask AsMarkdownReport(this SettingsTask settings) => 
        settings.UseExtension("md").AutoVerify(includeBuildServer: false);
    
    public static VerifySettings AsFreezeFile(this VerifySettings settings)
    {
        settings.UseExtension("freeze.txt");
        return settings;
    }

    public static VerifySettings AsMarkdownReport(this VerifySettings settings)
    {
        settings.UseExtension("md");
        settings.AutoVerify(includeBuildServer: false);
        return settings;
    }
}

```

and example use cases

```c#

    [Test]
    public async Task PermissionsMarkdownReport() =>
        await Verify(BuildPermissionsReport()).AsMarkdownReport();

    [Test]
    public async Task EfCore_entities_must_have_tenant_attribute() => 
        await Verify(MissingTenantAttribute(EntityTypes).Select(t => t.Name))
            .AsFreezeFile();
```

Does this give you enough context?

**Reply by drewburlingame** [reply]
@SimonCropp just checking in.  Were these examples enough to give you the context of what we're looking for?  If not, I'll carve out some time to put up a repo for you.   Was hoping to save us both some time if this was enough.

**Comment by SeanFeldman** [comment]
Same here. The code used to verify image types with the output filenames constructed using the extension of the input type. 

```
 public static IEnumerable<object[]> ImageTypePermutations()
{
            yield return new object[] { "jpeg", new JpegEncoder(), "image/jpeg"};
            yield return new object[] { "png", new PngEncoder(), "image/png"};
            yield return new object[] { "gif", new PngEncoder(), "image/gif"};
}
```

With `UseExtension()` gone, the output that's generated is all `.bin`

![image](https://github.com/VerifyTests/Verify/assets/1309622/de707d33-f8c7-4f82-b736-7c7cb341e2cb)

**Reply by SeanFeldman** [reply]
The fix is now to supply the extension in the `Verify()` call as an optional parameter.

```
await Verify(stream, extension: imageExtension)
```

In case you'll be reviewing the API, @SimonCropp, this is less intuitive than an extension method with a fluent API or `VerifySettings` parameter passed into `Verify()`.
